design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,121 @@
+
+### What is in scope
+
+* Syntax for defining lambda functions will be proposed
+* Lambda invocation functions will be proposed
+
+## What is not in scope
+
+* Implementation details will not be proposed
+* Level of effort will not be estimated
+
+## Value/return
+
+The introduction of lambda functions will ultimately empower ksqDB users to solve more problems with less effort. Many of the UDFs we’ve seen users write are designed to apply relatively simple functionality to the elements of `Arrays` and `Maps`. Lambda functions are well-suited to solve this class of problems in a very easy and intuitive way.
+
+Lambda functions will also enable some users to use ksqlDB in environments that do not allow for the deployment of UDFs.
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
+
+```
+arg => expr
+```
+
+This expression represents the following:
+
+* `arg` is the argument passed into the scope of each invocation of the lambda function.
+* `expr` is the "body" of the lambda function. Each invocation of the lambda function will evaluate this expression and return the result.
+
+Multiple arguments should also be allowed for lambda functions. When multiple arguments are used, they should be wrapped in parentheses:
+
+```
+(x, y) => expr
+```
+
+#### Allowed expressions
+
+The body of each lambda function should simply be **any expression that could be used in a `SELECT` expression list**, with some exceptions:
+
+* Aggregates should not be allowed (please see *[Open questions](#open-questions)* section)
+* `SELECT` subqueries should not be allowed
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_map(map, (k, v) => x, y)` - Applies the given lambda function to each key-value pair of the input `Map`, returning a new `Map` containing the transformed key-value pairs.
```

Comment:
Are we sure, that allowing multiple return values does not cause problems?

Reply:
I'm not super familiar with the internals but I am guessing that this would require quite a bit of effort since we don't really return tuples anywhere that I know of. Perhaps it would be worth separating `transform_map` out into `transform_keys` and `transform_values`? You could easily combine those to effectively get what `transform_map` would give you anyways:
```
transform_keys(transform_values(m, v => expr(v)), k => expr(k))
```
I think this approach might even be preferable, as in this case the lambda function is only applied to one "thing" at a time. It feels a little weird to apply a lambda function to a key and value simultaneously.

Reply:
I updated the proposal to use `transform_keys` and `transform_values` instead of `transform_map`.

Reply:
I don't know the code details either. Would need input from an kslqDB engineer on this.